site/src/pages/WorkspacePage/WorkspacePage.tsx

```diff
@@ -1,54 +1,100 @@
   const params = useParams() as {
     username: string;
     workspace: string;
   };
   const workspaceName = params.workspace;
   const username = params.username.replace("@", "");
   const orgId = useOrganizationId();
-  const [workspaceState, workspaceSend] = useMachine(workspaceMachine, {
-    context: {
-      orgId,
-      workspaceName,
-      username,
-    },
-    actions: {
-      refreshBuilds: async () => {
-        await buildsQuery.refetch();
-      },
-    },
+
+  // Workspace
+  const workspaceQueryOptions = workspaceByOwnerAndName(
+    username,
+    workspaceName,
+  );
+  const workspaceQuery = useQuery(workspaceQueryOptions);
+  const workspace = workspaceQuery.data;
+
+  // Template
+  const templateQuery = useQuery({
+    ...templateByName(orgId, workspace?.template_name ?? ""),
+    enabled: workspace !== undefined,
   });
-  const { workspace, error } = workspaceState.context;
-  const quotaQuery = useQuery(workspaceQuota(username));
-  const pageError = error ?? quotaQuery.error;
+  const template = templateQuery.data;
+
+  // Permissions
+  const checks =
+    workspace && template ? workspaceChecks(workspace, template) : {};
+  const permissionsQuery = useQuery({
+    ...checkAuthorization({ checks }),
+    enabled: workspace !== undefined && template !== undefined,
+  });
+  const permissions = permissionsQuery.data as WorkspacePermissions | undefined;
+
+  // Builds
   const buildsQuery = useInfiniteQuery({
     ...infiniteWorkspaceBuilds(workspace?.id ?? ""),
-    enabled: Boolean(workspace),
+    enabled: workspace !== undefined,
   });
 
-  const canAutostartResponse = useQuery(
-    workspaceResolveAutostart(workspace?.id ?? ""),
-  );
+  // Watch workspace changes
+  const workspaceEventSource = useRef<EventSource | null>(null);
+  useEffect(() => {
+    // If there is an event source, we are already watching the workspace
+    if (!workspace || workspaceEventSource.current) {
+      return;
+    }
+
+    const eventSource = watchWorkspace(workspace.id);
+    workspaceEventSource.current = eventSource;
 
-  const canAutostart = !canAutostartResponse.data?.parameter_mismatch ?? false;
+    eventSource.addEventListener("data", async (event) => {
+      const newWorkspaceData = JSON.parse(event.data) as Workspace;
+      queryClient.setQueryData(
+        workspaceQueryOptions.queryKey,
+        newWorkspaceData,
+      );
+
+      const hasNewBuild =
+        newWorkspaceData.latest_build.id !== workspace.latest_build.id;
+      const lastBuildHasChanged =
+        newWorkspaceData.latest_build.status !== workspace.latest_build.status;
+
+      if (hasNewBuild || lastBuildHasChanged) {
+        await buildsQuery.refetch();
+      }
+    });
+
+    eventSource.addEventListener("error", (event) => {
+      console.error("Error on getting workspace changes.", event);
+    });
+
+    return () => {
+      eventSource.close();
+    };
+  }, [buildsQuery, queryClient, workspace, workspaceQueryOptions.queryKey]);
```

Comment:
Can you go into more detail about when you want the effect to sync? Is it basically whenever a new workspace is available?

Because it looks like there might be a bug in the dependency array – `workspaceQueryOptions.queryKey` will always produce a new array by memory reference, which will make the effect's cleanup function get called the moment anything in the top-level component re-renders. `buildsQuery` might also have similar behavior – I don't know if React Query guarantees whether a query object maintains the same external reference across renders

I'm guessing you only really want the `close` method to get called when the workspace changes?

Reply:
Yes, exactly but I wanted to keep the dependency array with all the inside references just as the lint suggests but I think it is not a good idea at all. I'm wondering if I can just manually add the deps that I really want to sync and ignore the lint for this case.... wdyt?

Reply:
Maybe useEffectEvent might have a good use case here?

```ts
// event would need a type - not sure what it is off the top of my head
const onDataChange = useEffectEvent(async (event) => {
  const newWorkspaceData = JSON.parse(event.data) as Workspace;
  queryClient.setQueryData(
    workspaceQueryOptions.queryKey,
    newWorkspaceData,
  );

  const hasNewBuild =
    newWorkspaceData.latest_build.id !== workspace.latest_build.id;
  const lastBuildHasChanged =
    newWorkspaceData.latest_build.status !== workspace.latest_build.status;

  if (hasNewBuild || lastBuildHasChanged) {
    await buildsQuery.refetch();
  }
});

useEffect(() => {
  if (!workspace) {
    return;
  }
  
  const eventSource = watchWorkspace(workspace.id);
  eventSource.addEventListener("data", onDataChange);
  eventSource.addEventListener("error", (event) => {
    console.error("Error on getting workspace changes.", event);
  });
  
  return () => eventSource.close();
}, [onDataChange, workspace]);
```
And because the workspace is the only "cue" for re-syncing, that means you can get rid of the ref state, and move everything inside the effect

Reply:
Yes, I used it. Thank you!